File: dron_cinematico/scripts/lider_seguidor/lider_seguidor.py
#!/usr/bin/env python

#Use 'python' for ROS Kinetic and Melodic.
#Use 'python3' for ROS Noetic
"""This python script is the equivalent to the 'src/two_hector_quadrotors_lf.cpp', so it executes the same tasks.
For more information, please refers to that source file.

-- Important: To land the drones and finish this node, please press 'q' key."""

__author__ = "C. Mauricio Arteaga-Escamilla"

import rospy, math
from hector_uav_msgs.srv import EnableMotors #rosservice info /enable_motors 
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Twist
from tf.transformations import euler_from_quaternion, quaternion_from_euler
from math import sin, cos, pi

#kbhit function implemented on Linux
import sys, select, os #Handling command line arguments
if os.name == 'nt':
  import msvcrt
else:
  import tty, termios

#Define the robots' posture using lists [x,y,z,roll,pitch,yaw]
r_pose1 = [0,0,0,0,0,0]
r_pose2 = [0,0,0,0,0,0]
takeoff_alt = 0.5 #Desired initial altitude

vel1 = Twist() #Create Twist message instance
vel2 = Twist()

t = 0.0; t0 = 0.0; Vxy_max = 2;  Vz_max = 0.5; Wz_max = 4; #timer, initial time, maximum velocities [m/s, rad/s], respectively
T = 100.0; k = 0.1; #Trajectory period, controller gains kx = ky = k

#Leader drone variables
ex = 0.0; ey = 0.0; ez = 0; e_yaw = 0
Vxl = 0; Vyl = 0; Vzl = 0; Wzl = 0 #Leader control signals

#Follower drone variables
exf = 0; eyf = 0; ezf = 0; e_yawf = 0; s_d = 1.5; alp_d = pi; s_zd = 0 #Follower tracking errors and desired formation states
Vxf = 0; Vyf = 0; Vzf = 0; Wzf = 0 #Follower control signals


def getKey(): #Function to use keyboard events on Linux
    if os.name == 'nt':
      return msvcrt.getch()

    tty.setraw(sys.stdin.fileno())
    rlist, _, _ = select.select([sys.stdin], [], [], 0.1)
    if rlist:
        key = sys.stdin.read(1)
    else:
        key = ''

    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
    return key

def enable_motors(): #Function to call "/enable_motors" service, needed to move the drone
    #drone 1
    SERVICE_ENABLE_MOTORS = "uav1/enable_motors"
    print("Waiting for service", SERVICE_ENABLE_MOTORS)
    rospy.wait_for_service(SERVICE_ENABLE_MOTORS)
    try:
        enable_motors = rospy.ServiceProxy(SERVICE_ENABLE_MOTORS, EnableMotors)
        res = enable_motors(True)
        if res:
            print("uav1: Enable motors successful\n")
        else:
            print("uav1: Enable motors failed\n")
    except rospy.ServiceException:
        print("Enable service", SERVICE_ENABLE_MOTORS, "call failed")
    
    #drone 2
    SERVICE_ENABLE_MOTORS = "uav2/enable_motors"
    print("Waiting for service", SERVICE_ENABLE_MOTORS)
    rospy.wait_for_service(SERVICE_ENABLE_MOTORS)
    try:
        enable_motors = rospy.ServiceProxy(SERVICE_ENABLE_MOTORS, EnableMotors)
        res = enable_motors(True)
        if res:
            print("uav2: Enable motors successful\n")
        else:
            print("uav2: Enable motors failed\n")
    except rospy.ServiceException:
        print("Enable service", SERVICE_ENABLE_MOTORS, "call failed")

def poseCallback1(msg): #Callback function to get the drone 1 posture
        global r_pose1
        r_pose1[0] = msg.pose.pose.position.x
        r_pose1[1] = msg.pose.pose.position.y
        r_pose1[2] = msg.pose.pose.position.z
        
        #Operations to convert from quaternion to Euler angles
        quater = msg.pose.pose.orientation
        quater_list = [quater.x, quater.y, quater.z, quater.w]
        (r_pose1[3], r_pose1[4], r_pose1[5]) = euler_from_quaternion(quater_list) #Euler angles are given in radians

def poseCallback2(msg): #Callback function to get the drone 2 posture
	global r_pose2
	r_pose2[0] = msg.pose.pose.position.x
	r_pose2[1] = msg.pose.pose.position.y
	r_pose2[2] = msg.pose.pose.position.z
	
	#Operations to convert from quaternion to Euler angles
	quater = msg.pose.pose.orientation
	quater_list = [quater.x, quater.y, quater.z, quater.w]
	(r_pose2[3], r_pose2[4], r_pose2[5]) = euler_from_quaternion(quater_list)

def move1(x,y,z,turn): #Function to assign control signals (Vx, Vy, Vz, Wz)
       vel1.linear.x = x; vel1.linear.y = y
       vel1.linear.z = z; vel1.angular.z = turn
       vel_pub1.publish(vel1)

def move2(x,y,z,turn):
	vel2.linear.x = x; vel2.linear.y = y
	vel2.linear.z = z; vel2.angular.z = turn
	vel_pub2.publish(vel2)

def takeoff(): #Takeoff function of all drones
	rospy.logwarn(" Desired takeoff altitude = %.2f\n Taking off ...\n", takeoff_alt)
	
	#Quadrotor 1
	move1(0,0,0.3,0)	
	
	while(r_pose1[2] < takeoff_alt-0.1):
		vel_pub1.publish(vel1)
		rate.sleep(); rospy.sleep(0.1)
		key = getKey()
		if(key == 'q'):
			break
	
	move1(0,0,0,0) #Stop
	
	#Quadrotor 2
	move2(0,0,0.3,0)	
	
	while(r_pose2[2] < takeoff_alt-0.1): #Compare if the desired altitude is reached
		vel_pub2.publish(vel2)
		vel_pub1.publish(vel1) #Important: velocities of the drone 1 must keep being published to avoid a not desired landing
		rate.sleep(); rospy.sleep(0.1)
		key = getKey()
		if(key == 'q'):
			break
	
	move2(0,0,0,0) #Stop

def land(): #Land function of all drones
	print("\n Landing ...\n")
	
	move1(0,0,-0.2,0); move2(0,0,-0.2,0)
		
	while(r_pose1[2] > 0.3 or r_pose2[2] > 0.3):
		vel_pub1.publish(vel1)
		vel_pub2.publish(vel2)
		rate.sleep(); rospy.sleep(0.1)
	
	move1(0,0,0,0); move2(0,0,0,0)

def leader_vel_control(): #Function to generate the desired trajectory and to compute the signals control of the leader robot
	global ex, ey, ez, e_yaw, Vxl, Vyl, vzl, Wzl #Indicate that some variables are global to be used in the main_function
	#Desired trajectory: Helix
	X0 = 0; Y0 = 0; Z0 = 0.5; radius = 3; w = 2*pi/T; Vzd = 0.02

	#Desired position in the 3D space
	Xd = X0+radius*sin(w*t)
	Yd = Y0+radius*cos(w*t)
	Zd = Z0+Vzd*t #linear increasing
	Yawd = sin(w*t) #sinusoidal behavior
	
	#Corresponding time derivatives
	Xdp = radius*w*cos(w*t)
	Ydp = -radius*w*sin(w*t)
	Zdp = Vzd
	Yawdp = w*cos(w*t)
	
	
	ex = r_pose1[0]-Xd; ey = r_pose1[1]-Yd; ez = r_pose1[2]-Zd; e_yaw = r_pose1[5]-Yawd #Compute tracking errors
	
	#Cinematic controller. Auxiliar controls, in global coordinates (local variables)
	Ux = Xdp-k*ex; Uy = Ydp-k*ey
	
	#Translational velocities with respect to the robot frame
	Vxl = Ux*cos(r_pose1[5])+Uy*sin(r_pose1[5])
	Vyl = -Ux*sin(r_pose1[5])+Uy*cos(r_pose1[5])
	
	#Cinematic controller. Note: Vz and Wz are compute directly since they are given with respect to the robot frame
	Vzl = Zdp-k*ez
	Wzl = Yawdp-k*e_yaw

	#Velocities saturation
	if(abs(Vxl)>Vxy_max):
		Vxl = Vxy_max*abs(Vxl)/Vxl; print("Sat Vxl\t")
	if(abs(Vyl)>Vxy_max):
		Vyl = Vxy_max*abs(Vyl)/Vyl; print("Sat Vyl\t")
	if(abs(Vzl)>Vz_max):
		Vzl = Vz_max*abs(Vzl)/Vzl; print("Sat Vzl\t")
	if(abs(Wzl)>Wz_max):
		Wzl = Wz_max*abs(Wzl)/Wzl; print("Sat Wzl\t")
	
	move1(Vxl,Vyl,Vzl,Wzl) #Publish the 4 control signals

def follower_vel_control(): #Function to compute the signals control of the follower robot
	#Note: desired follower position and yaw orientation are given by the leader's ones
	global exf, eyf, ezf, e_yawf #Indicate that some variables are global to be used in the main_function
	
	#Follower desired position with respect to the gloabl frame
	Xfd = r_pose1[0]+s_d*cos(r_pose1[5]+alp_d)
	Yfd = r_pose1[1]+s_d*sin(r_pose1[5]+alp_d)
	
	#Translational velocities of the leader drone in the global frame
	xlp = Vxl*cos(r_pose1[5])-Vyl*sin(r_pose1[5])
	ylp = Vxl*sin(r_pose1[5])+Vyl*cos(r_pose1[5])

	#Corresponding desired time derivatives
	Xfdp = xlp-Wzl*s_d*sin(r_pose1[5]+alp_d)
	Yfdp = ylp+Wzl*s_d*cos(r_pose1[5]+alp_d)
	
	#Compute tracking errors (with respect to the global frame)
	exf = r_pose2[0]-Xfd; eyf = r_pose2[1]-Yfd
	ezf = r_pose2[2]-(r_pose1[2]+s_zd); e_yawf = r_pose2[5] - r_pose1[5] #Both the altitude and the orientation are equal to leader drone's ones
	
	#Cinematic controller. Auxiliar controls, in global coordinates (local variables)
	Ux = Xfdp-k*exf; Uy = Yfdp-k*eyf
	
	#Translational velocities with respect to the robot frame
	Vxf = Ux*cos(r_pose2[5])+Uy*sin(r_pose2[5])
	Vyf = -Ux*sin(r_pose2[5])+Uy*cos(r_pose2[5])
	
	#Vertical and rotational velocities
	Vzf = Vzl-k*ezf
	Wzf = Wzl-k*e_yawf

	#Velocities saturation
	if(abs(Vxf)>Vxy_max):
		Vxf = Vxy_max*abs(Vxf)/Vxf; print("Sat Vxf\t")
	if(abs(Vyf)>Vxy_max):
		Vyf = Vxy_max*abs(Vyf)/Vyf; print("Sat Vyf\t")
	if(abs(Vzf)>Vz_max):
		Vzf = Vz_max*abs(Vzf)/Vzf; print("Sat Vzf\t")
	if(abs(Wzf)>Wz_max):
		Wzf = Wz_max*abs(Wzf)/Wzf; print("Sat Wzf\t")
	
	move2(Vxf,Vyf,Vzf,Wzf) #Publish the 4 control signals


def main_function():
	rospy.init_node("two_hector_quadrotors_lf", anonymous=True) #Initialize the node. anonymous=True for multiple nodes
	global rate
	rate = rospy.Rate(50) #Node frequency (Hz)
	counter = 0
	
	global vel_pub1, vel_pub2
	vel_pub1 = rospy.Publisher('uav1/cmd_vel', Twist, queue_size=10) #To publish in the topic
	vel_pub2 = rospy.Publisher('uav2/cmd_vel', Twist, queue_size=10)
	rospy.Subscriber('uav1/ground_truth/state', Odometry, poseCallback1) #To subscribe to the topic
	rospy.Subscriber('uav2/ground_truth/state', Odometry, poseCallback2)
	
	
	enable_motors() #Execute the functions
	takeoff()
	
	print("Press 'q' to land the drones and finish the node\n")
	rospy.logwarn(" To start the movement, the simulation must be running\n\n") #Warning message
	
	global t, t0 #t and t0 are global to be used in leaer_vel_control()
	t0 = rospy.Time.now().to_sec()
	
	while(1):
		t = rospy.Time.now().to_sec()-t0
		leader_vel_control() #Compute the control signals of the leader robot
		follower_vel_control() #Compute the control signals of the follower robot
		
		if counter == 25: #Frequency divisor
			rospy.loginfo("ex: %.3f ey: %.3f ez: %.2f e_yaw: %.3f\n", ex,ey,ez,e_yaw)
			rospy.loginfo("exf: %.3f eyf: %.3f ezf: %.2f e_yawf: %.3f\n", exf,eyf,ezf,e_yawf)
			counter = 0 #Reset the counter
		else:
			counter += 1
		
		rate.sleep() #spinOnce() function does not exist in python
		key = getKey()
		if(key == 'q'):
			break

	land() #Execute land function
	print("\n Node finished\n")


if __name__ == '__main__':
    if os.name != 'nt':
        settings = termios.tcgetattr(sys.stdin)

    try:
    	main_function()  #Execute the function
	
    except rospy.ROSInterruptException:
        pass

    #if os.name != 'nt':
    #   termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)

File: dron_cinematico/src/lider_seguidor/lider_seguidor.cpp
/* 
@descripcion:

Este nodo ejecuta el controlador para resolver el problema del control de formacion usando el esquema lider-seguidor con 2 drones simulados en gazebo. El dron lider sigue su trayectoria deseada utilizando la funcion "velocity controller" y entonces el robot seguidor sigue la trayectoria desfasada del robot lider. 

primero se hace una llamda al servicio "/enable_motors" para cada uno d elos drones. Se definden las funciones de despegue y aterrizaje para ambos drones.

-- para aterrizar ambos drones y termianr este nodo, se necesita presioanr la tecla 'ESC'.

Instrucciones:
En diferentes terminales:
$ roslaunch my_hector_uavs two_hector_uavs.launch
	# Lanza un quadrotor simulado en Gazebo usando un mundo vacío. Nota: La simulación comienza en pausa.

$ rosrun my_hector_uavs two_hector_quadrotors_lf
	# Corre el nodo
*/

// se agregan los archivos de cabecera 
#include "ros/ros.h"
#include "hector_uav_msgs/EnableMotors.h" //Hacer la llamada al servicio --> roservice info/enable_motors
#include "nav_msgs/Odometry.h"
#include "geometry_msgs/Twist.h"
#include "tf/transform_broadcaster.h"
//los ultimos 3 se necesitan para utilizar eventos del teclado.
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

using namespace std;
// se crea una estructura personaliazda para definir la posicion y orientacion global del dron utilizando los angulos de euler
struct robot_pose{ double x, y, z, roll, pitch, yaw; };  

//Se definen objetos globales 
ros::Publisher  vel_pub1, vel_pub2; //Publicar velocidades
ros::Subscriber pose_sub1, pose_sub2; //Suscribirse a la odometria 
geometry_msgs::Twist vel1, vel2;
ros::ServiceClient enable_motors_client1, enable_motors_client2; // Hacer la llamda al servicio enble motors

int key; //Variable para guardar el valor de la clave
robot_pose r_pose1, r_pose2; //Definir variables globales (posición y orientación, usando mi estructura personalizada)
float takeoff_alt = 0.5; // Altitud inicial deseada

double t, t0, Vxy_max = 2, Vz_max = 0.5, Wz_max = 8; //Temporizador, tiempo inicial, velocidades de traslación máximas [m / s] y velocidades de rotación máximas [rad / s]

//Variables de drones líderes
double T = 100, k = 0.1; //Período de trayectoria, ganancia del controlador  kx = ky = k
double ex, ey, ez, e_yaw, Xd, Yd, Zd, Yawd, Xdp, Ydp, Zdp, Yawdp; //Seguimiento de errores, posiciones deseadas, orientación deseada y derivada del tiempo, respectivamente
double Vxl, Vyl, Vzl, Wzl; //señales de control para el lider 

//defino variables robot seguidor
double exf, eyf, ezf, e_yawf, s_d = 1.5, alp_d = M_PI, s_zd = 0; // error de seguimiento y estado de formacion deseado 
double Vxf, Vyf, Vzf, Wzf;  //Señales de control


int kbhit(void){ // funcion función de Windows implementada para Linux
  struct termios oldt, newt;
  int ch, oldf;

  tcgetattr(STDIN_FILENO, &oldt);
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
  fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

  ch = getchar();

  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
  fcntl(STDIN_FILENO, F_SETFL, oldf);

  if(ch != EOF)
  {
    ungetc(ch, stdin);
    return 1;
  }
  return 0;
}

void enable_motors(){ //Se llama a la funcion "/enable_motors" ,servicio necesario para mover al dron
  ROS_WARN_ONCE("Calling the enable_motors services ...\n");
  hector_uav_msgs::EnableMotors enable_request1, enable_request2; // llamo a los dos sercicios enable para cada dron 

  enable_request1.request.enable = true;
  enable_motors_client1.call(enable_request1);
  if(enable_request1.response.success){ cout << "uav1: enable motors successful\n"; }
  else{ cout << "uav1: enable motors failed" << endl; }

  enable_request2.request.enable = true;
  enable_motors_client2.call(enable_request2);
  if(enable_request2.response.success){ cout << "uav2: enable motors successful\n"; }
  else{ cout << "uav2: enable motors failed" << endl; }
}

void poseCallback1(const nav_msgs::Odometry::ConstPtr msg){ //Función de devolución de llamada para obtener la postura del dron 1
  r_pose1.x = msg->pose.pose.position.x;
  r_pose1.y = msg->pose.pose.position.y;
  r_pose1.z = msg->pose.pose.position.z;
  //Operaciones para convertir de cuaternión a ángulos de Euler
  tf::Quaternion quater;
  tf::quaternionMsgToTF(msg->pose.pose.orientation, quater);
  tf::Matrix3x3(quater).getRPY(r_pose1.roll, r_pose1.pitch, r_pose1.yaw);  //Los ángulos de Euler se expresan en radianes.
}

void poseCallback2(const nav_msgs::Odometry::ConstPtr msg){ //Función de devolución de llamada para obtener la postura del dron 2
  r_pose2.x = msg->pose.pose.position.x;
  r_pose2.y = msg->pose.pose.position.y;
  r_pose2.z = msg->pose.pose.position.z;
  //Operaciones para convertir de cuaternión a ángulos de Euler
  tf::Quaternion quater;
  tf::quaternionMsgToTF(msg->pose.pose.orientation, quater);
  tf::Matrix3x3(quater).getRPY(r_pose2.roll, r_pose2.pitch, r_pose2.yaw);  //Los ángulos de Euler se expresan en radianes.
}

void move1(float x, float y, float z, float turn){ //Función para asignar señales de control y publicarlas, drone 1
  vel1.linear.x = x; vel1.linear.y = y; vel1.linear.z = z; vel1.angular.z = turn;
  vel_pub1.publish(vel1);
}
void move2(float x, float y, float z, float turn){ ///Función para asignar señales de control y publicarlas, drone 2
  vel2.linear.x = x; vel2.linear.y = y; vel2.linear.z = z; vel2.angular.z = turn;
  vel_pub2.publish(vel2);
}

void takeoff(){ //Función de despegue de todos los drones.
  cout << " Desired takeoff altitude = " << takeoff_alt << "\n Taking off ...\n";

  //Quadrotor 1
  move1(0,0,0.3,0); //Asigno velocidades de ascenso
  
  while(r_pose1.z < takeoff_alt-0.1){ //Compare si se alcanza la altitud deseada
    vel_pub1.publish(vel1);
    ros::spinOnce(); ros::Duration(0.1).sleep();
    if(kbhit())	key = getchar(); //Compruebe cualquier pulsación de tecla
    if(key == 27) break;
  }
  move1(0,0,0,0); //Detener

  //Quadrotor 2
  move2(0,0,0.3,0); //Asigno velocidades de ascenso

  while(r_pose2.z < takeoff_alt-0.1){ //Compare si se alcanza la altitud deseada
    vel_pub2.publish(vel2);
    vel_pub1.publish(vel1); // Importante: las velocidades del dron 1 deben seguir publicándose para evitar un aterrizaje no deseado
    ros::spinOnce(); ros::Duration(0.1).sleep();
    if(kbhit())	key = getchar(); //Compruebe cualquier pulsación de tecla
    if(key == 27) break;
  }
  move2(0,0,0,0); //detener
}

void land(){ //Funcion de aterrizaje de todos los drones 
  cout << "\n Landing ...\n";

  move1(0,0,-0.3,0); move2(0,0,-0.3,0); //Asigno velocidades de descenso

  while(r_pose1.z > 0.3 || r_pose2.z > 0.3){
    vel_pub1.publish(vel1);  vel_pub2.publish(vel2);
    ros::spinOnce(); ros::Duration(0.1).sleep();
  }
  move1(0,0,0,0);  move2(0,0,0,0); //Apaga todos los drones
}

void leader_vel_control(){ //Función para generar la trayectoria deseada y calcular el control de señales del robot líder
  // Definir la trayectoria deseada:
  double X0 = 0, Y0 = 0, Z0 = 0.5, radius = 3, w = 2*M_PI/T, Vzd = 0.02;

  //Posición deseada en el espacio 3D
  Xd = X0+radius*sin(w*t);
  Yd = Y0+radius*cos(w*t);
  Zd = Z0+Vzd*t; //aumento lineal
  Yawd = sin(w*t); //comportamiento sinusoidal

  //Derivadas de tiempo correspondientes
  Xdp = radius*w*cos(w*t);
  Ydp = -radius*w*sin(w*t);
  Zdp = Vzd;
  Yawdp = w*cos(w*t);


  ex = r_pose1.x-Xd; ey = r_pose1.y-Yd; ez = r_pose1.z-Zd; e_yaw = r_pose1.yaw - Yawd; //Calcular errores de seguimiento

  //Controlador cinematográfico. Controles auxiliares, en coordenadas globales (variables locales)
  double Ux = Xdp-k*ex; double Uy = Ydp-k*ey;
  
  //Velocidades de traslación con respecto a la estructura del robot
  Vxl = Ux*cos(r_pose1.yaw)+Uy*sin(r_pose1.yaw);
  Vyl = -Ux*sin(r_pose1.yaw)+Uy*cos(r_pose1.yaw);

  // Controlador cinematográfico. Nota: Vz y Wz se calculan directamente ya que se dan con respecto a la estructura del robot.
  Vzl = Zdp-k*ez;
  Wzl = Yawdp-k*e_yaw;

  //Velocidad de saturacion
  if(abs(Vxl)>Vxy_max){Vxl = Vxy_max*abs(Vxl)/Vxl; printf("Sat Vxl\t");}
  if(abs(Vyl)>Vxy_max){Vyl = Vxy_max*abs(Vyl)/Vyl; printf("Sat Vyl\t");}
  if(abs(Vzl)>Vz_max){Vzl = Vz_max*abs(Vzl)/Vzl; printf("Sat Vzl\t");}
  if(abs(Wzl)>Wz_max){Wzl = Wz_max*abs(Wzl)/Wzl; printf("Sat Wzl\t");}

  move1(Vxl,Vyl,Vzl,Wzl); //Publica las 4 señales de control
}

void follower_vel_control(){ // Función para calcular el control de señales del robot seguidor
  //Nota: la posición deseada del seguidor y la orientación de guiñada vienen dadas por los del líder.
  
  double Xfd, Yfd, Xfdp, Yfdp, xlp, ylp, Ux, Uy; //Crea algunas variables locales

  //Seguidor de la posición deseada con respecto al marco global
  Xfd = r_pose1.x+s_d*cos(r_pose1.yaw+alp_d);
  Yfd = r_pose1.y+s_d*sin(r_pose1.yaw+alp_d);

  //Velocidades de traslación del dron líder en el marco global
  xlp = Vxl*cos(r_pose1.yaw)-Vyl*sin(r_pose1.yaw);
  ylp = Vxl*sin(r_pose1.yaw)+Vyl*cos(r_pose1.yaw);

  //Derivadas de tiempo deseadas correspondientes
  Xfdp = xlp-Wzl*s_d*sin(r_pose1.yaw+alp_d);
  Yfdp = ylp+Wzl*s_d*cos(r_pose1.yaw+alp_d);

  //Calcular los errores de seguimiento (con respecto al marco global)
  exf = r_pose2.x-Xfd; 			eyf = r_pose2.y-Yfd; 
  ezf = r_pose2.z-(r_pose1.z+s_zd); 	e_yawf = r_pose2.yaw - r_pose1.yaw; //Tanto la altitud como la orientación son iguales a las del drone líder.
  
  //Controlador cinematico. Controles auxiliares, en coordenadas globales
  Ux = Xfdp-k*exf; 	Uy = Yfdp-k*eyf;
  
  // Velocidades de traslación con respecto a la estructura del robot
  Vxf = Ux*cos(r_pose2.yaw)+Uy*sin(r_pose2.yaw);
  Vyf = -Ux*sin(r_pose2.yaw)+Uy*cos(r_pose2.yaw);

  //Velocidades verticales y de rotación
  Vzf = Vzl-k*ezf;
  Wzf = Wzl-k*e_yawf;

  //Velocidad de saturacion
  if(abs(Vxf)>Vxy_max){Vxf = Vxy_max*abs(Vxf)/Vxf; printf("Sat Vxf\t");}
  if(abs(Vyf)>Vxy_max){Vyf = Vxy_max*abs(Vyf)/Vyf; printf("Sat Vyf\t");}
  if(abs(Vzf)>Vz_max){Vzf = Vz_max*abs(Vzf)/Vzf; printf("Sat Vzf\t");}
  if(abs(Wzf)>Wz_max){Wzf = Wz_max*abs(Wzf)/Wzf; printf("Sat Wzf\t");}

  move2(Vxf,Vyf,Vzf,Wzf); //Publica las 4 señales de control
}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "two_hector_quadrotors_lf");  //Inicializo el nodo
  ros::NodeHandle n; //Creo el identificador del nodo
  int freq = 50; //frecuencia del nodo (Hz)
  ros::Rate loop_rate(freq);
  int counter = 0;
  bool finish = false;

  enable_motors_client1 = n.serviceClient<hector_uav_msgs::EnableMotors>("uav1/enable_motors"); //llamo al servicio
  enable_motors_client2 = n.serviceClient<hector_uav_msgs::EnableMotors>("uav2/enable_motors");
  vel_pub1 = n.advertise<geometry_msgs::Twist>("uav1/cmd_vel", 1); //publicar en el topico
  vel_pub2 = n.advertise<geometry_msgs::Twist>("uav2/cmd_vel", 1);
  pose_sub1 = n.subscribe("uav1/ground_truth/state", 1, poseCallback1); //suscribirse al  topic
  pose_sub2 = n.subscribe("uav2/ground_truth/state", 1, poseCallback2);

  enable_motors(); //Ejecute las funciones "enable_motors" y "takeoff" (para todos los robots)

  takeoff();
 
  printf(" Press 'ESC' to land the drones and finish the node\n");
  ROS_WARN_ONCE(" To start the movement, the simulation must be running\n\n"); //Mensaje de advertencia


  t0 = ros::Time::now().toSec(); //Obtener el tiempo inicial

  do{
    t = ros::Time::now().toSec()-t0; //Calculo el tiempo del controlador     
    leader_vel_control(); //Calculo las señales de control del robot lider
    follower_vel_control(); //Calculo las señales de control del robot seguidor

    if(counter == 100){ //Divisior de frecuencia para mostrar los errores de seguimiento de forma mas lenta 
      printf("\nex: %.3f\tey: %.3f\tez: %.2f\te_yaw: %.2f\n", ex,ey,ez,e_yaw); // Imprimier en la terminal los errores de seguimiento [m] y el error de orientacion [rad]
      printf("exf: %.3f\teyf: %.3f\tezf: %.2f\te_yawf: %.2f\n", exf,eyf,ezf,e_yawf); //Errores de formación del seguidor de impresión y error de orientación
      
      counter = 0; // Reseteo el contador
    }else counter++;

    ros::spinOnce();    //Requerido para recibir funciones de devolución de llamada (call back)
    loop_rate.sleep();  //Comando para esperar el resto del tiempo para completar la frecuencia del bucle
    if(kbhit())	key = getchar();

    if(key == 27) finish = true;
  }while(finish == false); //Note: No se utiliza ros::ok() porque es necesario seguir utilizando funciones de ros 
  
  land(); //Funcion aterrizar el dron 
  printf("\n Node finished\n");
  return 0;
}